Bottom: 3736927f2f621ae83d4830280ac3a3ab93274c06
Top:    ca4ae90d6b4e7312717901f00c11daf8790d0b7a
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-07-14 20:57:46 +0800

vhost_net: honor batched dequeuing for busy polling

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index e3d7ea1..82ec6f7 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -601,10 +601,13 @@ static int peek_head_len(struct vhost_net_virtqueue *rvq, struct sock *sk)
 	return len;
 }
 
-static int sk_has_rx_data(struct sock *sk)
+static int sk_has_rx_data(struct vhost_net_virtqueue *vq, struct sock *sk)
 {
 	struct socket *sock = sk->sk_socket;
 
+	if (vq->rx_array)
+		return vhost_net_buf_peek(vq);
+
 	if (sock->ops->peek_len)
 		return sock->ops->peek_len(sock);
 
@@ -628,7 +631,7 @@ static int vhost_net_rx_peek_head_len(struct vhost_net *net, struct sock *sk)
 		endtime = busy_clock() + vq->busyloop_timeout;
 
 		while (vhost_can_busy_poll(&net->dev, endtime) &&
-		       !sk_has_rx_data(sk) &&
+		       !sk_has_rx_data(rvq, sk) &&
 		       vhost_vq_avail_empty(&net->dev, vq))
 			cpu_relax();
